File: Ejemplos/43-DockerAndTestContainers/src/main/java/dev/joseluisgs/services/PersonasServiceImpl.java
package dev.joseluisgs.services;

import dev.joseluisgs.errors.PersonaError;
import dev.joseluisgs.errors.PersonaNotFound;
import dev.joseluisgs.model.Persona;
import dev.joseluisgs.repository.PersonasRepository;
import io.vavr.control.Either;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class PersonasServiceImpl implements PersonasService {
    private final PersonasRepository personasRepository;
    private final Logger logger = LoggerFactory.getLogger(PersonasServiceImpl.class);

    public PersonasServiceImpl(PersonasRepository personasRepository) {
        this.personasRepository = personasRepository;
    }

    @Override
    public Either<PersonaError, List<Persona>> getAll() {
        logger.info("Obteniendo personas...");
        return Either.right(personasRepository.getAll());
    }

    @Override
    public Either<PersonaError, Persona> getById(int id) {
        logger.info("Obteniendo persona por id...");
        try {
            var personaOptional = personasRepository.getById(id);
            if (personaOptional.isPresent()) {
                return Either.right(personaOptional.get());
            } else {
                return Either.left(new PersonaNotFound("Persona no encontrada con id: " + id));
            }
        } catch (Exception e) {
            return Either.left(new PersonaError("Error al obtener persona por id: " + id));
        }

    }

    @Override
    public Either<PersonaError, Persona> create(Persona persona) {
        logger.info("Creando persona...");
        try {
            return Either.right(personasRepository.create(persona));
        } catch (Exception e) {
            return Either.left(new PersonaError("Error al crear persona"));
        }
    }

    @Override
    public Either<PersonaError, Persona> update(long id, Persona persona) {
        logger.info("Actualizando persona...");
        // La busco
        try {
            if (personasRepository.getById(id).isPresent()) {
                return Either.right(personasRepository.update(id, persona));
            } else {
                return Either.left(new PersonaNotFound("Persona no encontrada con id: " + id));
            }
        } catch (Exception e) {
            return Either.left(new PersonaError("Error al actualizar persona con id: " + id));
        }
    }

    @Override
    public Either<PersonaError, Persona> delete(long id) {
        logger.info("Eliminando persona...");
        // La busco
        try {
            var personaOptional = personasRepository.getById(id);
            if (personaOptional.isPresent()) {
                personasRepository.delete(id);
                return Either.right(personaOptional.get());

            } else {
                return Either.left(new PersonaNotFound("Persona no encontrada con id: " + id));
            }
        } catch (Exception e) {
            return Either.left(new PersonaError("Error al eliminar persona con id: " + id));
        }
    }

}

File: Ejemplos/33-EjemplosTestRepositorioServicio/src/main/kotlin/personas/services/PersonasServiceImpl.kt
package dev.joseluisgs.personas.services

import dev.joseluisgs.personas.exceptions.PersonaException
import dev.joseluisgs.personas.model.Persona
import dev.joseluisgs.personas.repository.PersonasRepository
import dev.joseluisgs.personas.services.cache.CachePersonas
import org.lighthousegames.logging.logging
import java.util.*

class PersonasServiceImpl(
    private val cache: CachePersonas,
    private val repository: PersonasRepository
) : PersonasService {
    private val logger = logging()
    override fun getAll(): List<Persona> {
        logger.debug { "Obteniendo del repositorio" }
        val personas = repository.getAll()
        logger.info { "Personas obtenidas: $personas" }
        return personas
    }

    override fun findByName(name: String): List<Persona> {
        logger.debug { "Buscando por nombre en el repositorio" }
        val personas = repository.findByName(name)
        logger.info { "Personas obtenidas: $personas" }
        return personas
    }

    override fun getById(id: UUID): Persona {
        // Primero buscamos en cache
        logger.debug { "Buscando por ID en cache" }
        val persona = cache.get(id)
        if (persona != null) {
            logger.info { "Persona encontrada en cache: $persona" }
            return persona
        }
        // Si no está en cache, buscamos en el repositorio
        logger.debug { "Buscando por ID en el repositorio" }
        val personaFromDb = repository.getById(id)
        if (personaFromDb != null) {
            logger.info { "Persona encontrada en el repositorio: $personaFromDb" }
            // Actualizamos el cache
            cache.put(personaFromDb.ID, personaFromDb)
            return personaFromDb
        }
        logger.error { "persona no encontrada con ID: $id" }
        throw PersonaException.PersonaNotFoundException(id.toString())
    }

    override fun save(persona: Persona): Persona {
        logger.debug { "Guardando en el repositorio persona $persona" }
        val result = repository.save(persona)
        logger.info { "Persona guardada en el repositorio: $result" }
        return result
    }

    override fun update(id: UUID, persona: Persona): Persona {
        logger.debug { "Actualizando en el repositorio con id: $id y datos: $persona" }
        // A la base de datos, eliminamos el persona antiguo del cache
        cache.remove(id)
        val result = repository.update(id, persona)
        if (result == null) {
            logger.error { "Persona no encontrada con ID: $id" }
            throw PersonaException.PersonaNotUpdatedException(id.toString())
        }
        logger.info { "Persona actualizada en el repositorio: $result" }
        return result
    }

    override fun delete(id: UUID): Persona {
        logger.debug { "Eliminando en el repositorio con id: $id" }
        // A la base de datos, eliminamos el persona antiguo del cache
        cache.remove(id)
        val result = repository.delete(id)
        if (result == null) {
            logger.error { "Persona no encontrada con ID: $id" }
            throw PersonaException.PersonaNotDeletedException(id.toString())
        }
        logger.info { "Persona eliminada en el repositorio: $result" }
        return result
    }
}